content/blog/2022-05-19-fast-brili.md

```diff
@@ -0,0 +1,141 @@
+implement an interpreter for Bril whose sole focus is speed.
+
+## Design overview
+
+### Language choice
+We chose C for an implementation language as it is known for its speed as well
+as simplicity, which means we won't inadvertently write code that is more
+expensive than necessary.
+
+### Big Red Bytecode
+In order to make our interpreter as efficient as possible, we have designed a
+very compressed representation of bril instructions: (almost) all of them are
+represented in 64 bits, with some taking multiple 64 bit words as needed. See
+the [documentation](https://github.coecis.cornell.edu/cs897/fast-bril/blob/master/doc/brb.pdf).
+
+In order to appease the C typechecker and not rely on undefined behavior, we use
+a union type to internally represent instructions, but most have the first 16
+bits as the opcode, so we can pull this out and then determine how to decode the
+rest of the instruction. In principle, we could also use `uint64_t` or
+`uint16_t[4]`, but using unions is the most user friendly, and lets the C
+compiler do more work to access the fields of structs, in case there are any
+cool tricks for that behind the scenes.
+
+### Features
+We support a superset of the behavior provided by `brili`, so options like `-p`
+work exactly the same. We also support the following:
+ - `-b` will read in bytecode instead of the Bril json
+ - `-bo <file>` will output the bytecode to `<file>`
+ - `-pr` will print the program to standard out (probably more useful with the
+         `-b` option)
+ - `-ni` will NOT run the interpreter.
+ - `-e <file>` will emit assembly to `<file>`.
+
+### The Assembly situation
+We translate temporaries to consecutive natural numbers (the first seen is 0,
+then 1, etc.), and we couldn't help but notice that these numbers could easily
+be multiplied by 8, and then just turned into stack locations. Also, there isn't
+currently a true compiler for Bril all the way to machine code, so we may as
+well add one.
+
+We have a working compiler to armv8 with no register allocation, and are working
+on linear scan register allocation in arm as well as x86 code
+generation. Unfortunately, these features aren't quite finished due to being
+undergraduate students with other finals, however there's a good chance they'll
+be fixed after the class is over (in the interest of having something nice to
+merge into the bril repo).
+
+Our assembly does depend on the C standard library for functions like `malloc`
+(to support the memory extension) as well as a few printing functions as doing
+it all with system calls just didn't seem like a great use of time. We have some
+functions in the `lib/` directory for printing base types, as well as parsing
+them in order for the `main` function to read command line arguments. A future
+improvement would be putting `malloc` and other C functions we call into this
+file so that we could keep a compiled version around and not have to link with
+C.
```

Comment:
This whole compiler is very impressive in that it works, but this section is pretty low on details about how it actually works. Could you maybe include an example or two of how the instructions are translated, or some other way to give people insights into what was hard/interesting/reusable? The general idea of turning a bytecode interpreter into a compiler is very important to the world more broadly, so elucidating the insight with a bit more crunchy technical detail could be very useful!